src/RateLimiter/Counter.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Web\RateLimiter;
+
+use Psr\SimpleCache\CacheInterface;
+
+/**
+ * Counter implements generic сell rate limit algorithm (GCRA) that ensures that after reaching the limit futher
+ * increments are distributed equally.
+ *
+ * @link https://en.wikipedia.org/wiki/Generic_cell_rate_algorithm
+ */
+final class Counter implements CounterInterface
+{
+    public const ID_PREFIX = 'rate-limiter-';
+
+    private const MILLISECONDS_PER_SECOND = 1000;
+
+    /**
+     * @var int period to apply limit to, in milliseconds
+     */
+    private int $period;
+
+    private int $limit;
+
+    /**
+     * @var float maximum interval before next increment, in milliseconds
+     * In GCRA it is known as emission interval.
+     */
+    private float $incrementInterval;
+
+    private ?string $id = null;
```

Comment:
Why it nullable? Can we pass value from constructor instead setter?

Reply:
https://github.com/yiisoft/yii-web/pull/204/files#diff-b529e2c30fdfe6601a69f5e3f7f63dafR83

if null then generated from request automatically

Reply:
1. To have it in the interface.
2. To allow it to be filled automatically by middleware if it's not set.